File: elements-of-programming-interviews/python/directory_path_normalization.py
from test_framework import generic_test


"""
Allowed chars:
- . (current path)
- .. (parent dir)
- / (subdir)
- alphanum (dir name)
- empty (same as .)

cases:
    - empty
    - /////// -> / 
    - / -> / 
    - /../.. -> /
    - ../.. -> ../..
    - derp -> derp
    - /derp -> /derp
    - /derp/.. -> /
    - derp/.. -> empty? 
    - derp/../.. -> ..

- Paths can be absolute (begin at '/') or relative (begin at current dir)
- for abspaths, / is highest parent; any '..' from '/' yields '/'
- for relpaths, each '..' either pops the last known dir, or appends '..' 
"""


def shortest_equivalent_path_first(path: str) -> str:
    """
    - is absolute? (first char)
    - split path on '/' 
    - keep stack
    - for each item in path list:
        - if char is ., skip it
        - if item is alphanum, push it
        - if item is ..:
            - if dir is absolute, pop if stack is nonempty, else skip
            - if dir is rel, pop if stack is nonempty and stack top not .., else push  
    - return joined stack; prepend '/' is absolute path
    """
    if not path:
        return path
    absolute = path[0] == '/'

    stack = []
    dirs = [item for item in path.split('/') if item]
    for item in dirs:
        if item == "..":
            if absolute and stack:
                stack.pop()
            elif not absolute and (not stack or stack[-1] == ".."):
                stack.append(item)
            elif not absolute and stack and stack[-1] != "..":
                stack.pop()
        elif item == ".":
            continue
        else:
            stack.append(item)

    return ("/" if absolute else "") + "/".join(stack)


def shortest_equivalent_path(path: str) -> str:
    if not path:
        return path

    stack = []
    for item in (item for item in path.split('/') if (item and item != ".")):
        if item == "..":
            if stack and (path[0] == '/' or stack[-1] != ".."):
                stack.pop()
            else:
                stack.append(item)
        else:
            stack.append(item)

    return ("/" if path.startswith('/') else "") + "/".join(stack)


if __name__ == '__main__':
    cases = [
        ("", ""),
        ("///////", "/"),
        ("/", "/"),
        ("/../..", "/"),
        ("../..", "../.."),
        ("derp", "derp"),
        ("/derp", "/derp"),
        ("/derp/..", "/"),
        ("derp/..", ""),
        ("derp/../..", "..")
    ]
    for path, expected in cases:
        actual = shortest_equivalent_path(path)
        assert actual == expected, f"{path}: {actual} != {expected}"

    exit(
        generic_test.generic_test_main('directory_path_normalization.py',
                                       'directory_path_normalization.tsv',
                                       shortest_equivalent_path))

File: elements-of-programming-interviews/2023/cpp/directory_path_normalization.cc
#include <string>
#include <vector>

#include "test_framework/generic_test.h"
using std::string;
using std::vector;

/*
- Cases:
  - absolute path (begins with /)
  - relative path (no prefix /)

- Tokens:
  - ".."
    - relative: if the top of the (nonempty) stack is not "..", should pop.
Otherwise, push it.
    - nonrelative: always pop the top of the (nonempty) stack
  - "."
    - represents current dir; can be ignored
  - ""
    - represents "//"; can be ignored (or treated as ".")
  - (any chars)
    - represents a directory name. Has to be pushed, may be removed.
*/

string assembleFinalPath(const vector<string>& dirs, bool isAbsolute) {
  string path = isAbsolute ? "/" : "";
  for (string dir : dirs) {
    path += dir + "/";
  }
  if (path.size() > 1) {
    path.erase(path.end() - 1);
  }
  return path;
}

void processAbsolute(vector<string>& dirs, vector<string>& processedDirs) {
  for (string dir : dirs) {
    if (dir == "..") {
      if (!processedDirs.empty()) {
        processedDirs.pop_back();
      }
    } else {
      processedDirs.push_back(dir);
    }
  }
}

void processRelative(vector<string>& dirs, vector<string>& processedDirs) {
  for (string dir : dirs) {
    if (dir == ".." && !processedDirs.empty() && processedDirs.back() != "..") {
      processedDirs.pop_back();
    } else {
      processedDirs.push_back(dir);
    }
  }
}

void stripPath(const string& path, vector<string>& dirs) {
  string curr = "";
  for (int i = 0; i <= path.size(); i++) {
    if ((i == path.size() || path[i] == '/')) {
      if (!curr.empty() && curr != ".") {
        dirs.push_back(curr);
      }
      curr.clear();
    } else {
      curr += path[i];
    }
  }
}

string ShortestEquivalentPath(const string& path) {
  if (path.empty()) {
    return path;
  }
  bool isAbsolute = (path[0] == '/');
  vector<string> dirs, processedDirs;
  stripPath(path, dirs);
  isAbsolute ? processAbsolute(dirs, processedDirs)
             : processRelative(dirs, processedDirs);
  return assembleFinalPath(processedDirs, isAbsolute);
}

int main(int argc, char* argv[]) {
  /*
  ShortestEquivalentPath("");
  ShortestEquivalentPath("home");
  ShortestEquivalentPath("home/foobar");
  ShortestEquivalentPath("../home/foobar");
  ShortestEquivalentPath("../../home/foobar");
  ShortestEquivalentPath("../../home/../foobar");
  ShortestEquivalentPath("/");
  ShortestEquivalentPath("/../../");
  ShortestEquivalentPath("/home");
  ShortestEquivalentPath("/home/foobar/");
  ShortestEquivalentPath("/../home/foobar/");
  ShortestEquivalentPath("/../../home/foobar");
  ShortestEquivalentPath("/../../home/../foobar");
  ShortestEquivalentPath("/.");
  */

  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"path"};
  return GenericTestMain(args, "directory_path_normalization.cc",
                         "directory_path_normalization.tsv",
                         &ShortestEquivalentPath, DefaultComparator{},
                         param_names);
}
